File: blind 75 leetcode/Arrays and Strings/maximum_product_subarray.py
'''
Given an integer array nums, find a contiguous non-empty subarray within the array that has the largest product, and return the product.

It is guaranteed that the answer will fit in a 32-bit integer.

A subarray is a contiguous subsequence of the array.
https://www.youtube.com/watch?v=lXVy6YWFcRM&ab_channel=RandomCoderRandomCoder
https://leetcode.com/problems/maximum-product-subarray

O(N) time (loops through the array once)
O(1) space (only using two variables, no arrays or extra contiguous memory)
'''

class Solution2:

	# revisited this problem on 7-4-2023 using
	# https://www.youtube.com/watch?v=hnswaLJvr6g&ab_channel=takeUforward
	def maxProduct(self, nums: List[int]) -> int:
        """
        keep track of the running product coming from the front (prefix), and from the back (suffix)
        the logic is that this running prefix and suffix product will capture all possible subarrays
        that can be found, so we just need to find the max between the prefix and suffix product during
        the iteration
        For example:
        2, 5, -1, 3, 6
        if we take the prefix starting at 2
        and suffix starting at 6
        prefix = 2 * 5 = 10
        suffix = 6 * 3 = 18
        greatest is 18
        
        prefix = 2 * 5 * -1 = -10
        suffix = 6 * 3 * -1 = -18
        greatest is still 18
        
        prefix = 2 * 5 * -1 * 3 = -30
        suffix = 6 * 3 * -1 * 5 = -90
        greatest is still 18
        
        prefix = 2 * 5 * -1 * 3 * 6 = -180
        suffix = 6 * 3 * -1 * 5 * 2 = -180
        greatest is still 18
        
        If any of the numbers are zero, we need to "restart" the prefix/suffix product to 1 since
        if we kept it at 0, all the subsequent products would become 0 as well.
        """
        greatest = float("-inf") 
        prefix = 1
        suffix = 1
        length = len(nums)
        for i in range(length):
            if prefix == 0:
                prefix = 1
            if suffix == 0:
                suffix = 1
            prefix = prefix * nums[i]
            suffix = suffix * nums[length - 1 - i]
            greatest = max(greatest, max(prefix, suffix))
        return greatest


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ## we need to set the result to max(nums) to handle an edge case where you have only one number in the nums
        ## like [-1], in this case we can't just set res to zero, otherwise we would get 0 as the maximum
        res = max(nums)
        
        ## we set the curMin and curMax to 1 since multiplying any number by 1 is just the same number
        curMin, curMax = 1, 1
        for n in nums:
            ## in the case num with value of 0, since any num multiplied by zero is zero, we just reset the curMin and curMax
            ## to 1 and move onto the next number, as if we skip the 0 entirely and "reset" the curMin and curMax
            if (n == 0):
                curMin, curMax = 1, 1
                continue
            tmp = n * curMax
            ## we have the tmp variable so that in the calculation of curMin, we will not use
            ## the new updated value of curMax
            
            ## the idea behind finding both the curMax and curMin is to deal with cases where we have negative and positive numbers mixed in
            ## Some examples:
            ## #1 (when n helps us find the max)
            ## if our array was [-1, 8] and n was 8
            ## the curMax right now would be -1
            ## if we were to calculate it, n * curMax = -8, n * curMin = -8, but n = 8. Meaning the curMax would be 8 

            ## #2 (when curMin helps us find the max)
            ## for an array like so [-1, -2, -3] and n was -3
            ## curMax is 2
            ## curMin is -2
            ## here, n * curMax = -6, n * curMin = 6, and n = -3, so curMin actually helps us find the proper max

            ## curMin would be the same process but using min instead of Max function.
            ## see the youtube link above for more explanation
            curMax = max(n * curMax, n * curMin, n)
            curMin = min(tmp, n * curMin, n)

            ## res would just be the max between the current res value and the curMax that we found
            res = max(res, curMax)
        return res

File: node_js_ds_and_algos_implementations/1D dynamic programming/maximum_product_subarray.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    maxProduct(nums) {
        /* 
        9/10/2026
        Samuel Yu's solution:
        https://neetcode.io/problems/maximum-product-subarray/discuss/qrPlODVjWMumaBKFkVIf

        Two pass by getting the running product from left to right and finding the max
        so far, resetting whenever we get nums[i] = 0 to avoid making the rest of the 
        product 0
        and then comparing it to the running product from right to left

        Reason for comparing both:
        If there are an odd count of negative numbers, this ensures that we handle the case
        when we multiply two negatives and it becomes positive
        for example:
        [1,-2,-3,-4]
        1 * -2 = -2, -2 * -3 = 6, 6 * -4 = -24
        whereas
        -4 * -3 = 12, 12 * -2 = -24, -24 * 1 = -24

        You can see from left to right, the max product is 6
        but from right to left, the max product is actually 12

        Example edge case with 0
        [1,2,0,4] 
        1 * 2 = 2
        2 * 0 = 0, but instead of resetting to 0, we reset to 1,
        otherwise, every remaining product will also be 0
        1 * 4 = 4, so we get a max product of 4
        */

        let res = nums[0]
        let prod = 1
        for (let i = 0; i < nums.length; ++i){
            prod = prod * nums[i]
            // catch edge case in JS where if you multiply a negative number with 0,
            // you get -0, so you need to turn it back to 0
            if (prod === -0){
                prod = 0 
            }
            res = Math.max(res, prod)
            if (nums[i] === 0){
                prod = 1   
            }
        }
        prod = 1
        for (let i = nums.length - 1; i >= 0; --i){
            prod = prod * nums[i]
            if (prod === -0){
                prod = 0 
            }
            res = Math.max(res, prod)
            if (nums[i] === 0){
                prod = 1
            }
        }
        return res
    }
}

/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    /*
    Prefix and Suffix Multiplication
    1) get the prefix multiplication by taking the cumulative product from front to back. Each cumuluative
    product represents the subarray's product up to that point
    2) get the suffix multiplication by taking the cumulative product from back to front. Each cumulative
    product represents the subarray's product up to that point
    3) Comparing the prefix (starting from front) and suffix (starting from back), take the max between each,
    as this will get the max product that is possible between the two subarrays.

    There's an edge case where the subarray's product becomes 0 (when calculating either prefix or suffix),
    to handle this you need to "reset" the subarray to be equal to just the number itself.

    Time: O(N)
    Space: O(N)

    It's possible to optimize the space by not tracking the prefix and suffix and just storing the products
    directly as a variable. See the python solution (maximum_product_subarrays.py)
    */
    let prefix = [...nums]
    let suffix = [...nums]
    for (let i = 1; i < nums.length; ++i){
        if (prefix[i-1] === 0){
            prefix[i] = nums[i]
        }
        else {
            prefix[i] = nums[i] * prefix[i-1]
        }
    }
    for (let i = nums.length-2; i >= 0; --i){
        if (suffix[i+1] === 0){
            suffix[i] = nums[i]

        }
        else{
            suffix[i] = nums[i] * suffix[i+1]
        }
    }
    let max = Number.NEGATIVE_INFINITY
    for (let i = 0; i < nums.length; ++i){
        j = nums.length - 1 - i
        max = Math.max(max, Math.max(prefix[i], suffix[j]))
    }
    return max
    
};